flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/NettyShuffleUtils.java

```diff
@@ -119,6 +142,93 @@ public static int computeNetworkBuffersForAnnouncing(
         return requirementForInputs + requirementForOutputs;
     }
 
+    public static int maxRequiredBuffersPerGate(
+            ResultPartitionType partitionType, Optional<Integer> requiredMaxBuffersPerGate) {
+        int requiredMaxBuffers =
+                requiredMaxBuffersPerGate.orElseGet(
+                        () ->
+                                isPipelineResultPartition(partitionType)
+                                        ? DEFAULT_MAX_BUFFERS_PER_GATE_FOR_STREAM
+                                        : DEFAULT_MAX_BUFFERS_PER_GATE_FOR_BATCH);
+        checkState(requiredMaxBuffers >= 0, "Max required buffers per gate must be non-negative.");
+        return requiredMaxBuffers;
+    }
+
+    private static int getNumBuffersToAnnounceForInputGates(
+            ResultPartitionType type,
+            int configuredNetworkBuffersPerChannel,
+            int floatingNetworkBuffersPerGate,
+            Optional<Integer> requiredMaxBuffersPerGate,
+            int numInputChannels) {
+        int maxGateBuffersThreshold = maxRequiredBuffersPerGate(type, requiredMaxBuffersPerGate);
+        int expectMaxGateBuffers =
+                getExpectMaxBuffersPerGate(
+                        numInputChannels,
+                        configuredNetworkBuffersPerChannel,
+                        floatingNetworkBuffersPerGate);
+
+        int actualExclusiveBuffersPerChannel =
+                adjustExclusiveBuffersPerChannel(
+                        configuredNetworkBuffersPerChannel,
+                        numInputChannels,
+                        maxGateBuffersThreshold);
+        int actualExclusiveBuffersPerGate =
+                getActualExclusiveBuffersInGate(numInputChannels, actualExclusiveBuffersPerChannel);
+        int maxFloatingBuffers = Math.max(1, expectMaxGateBuffers - actualExclusiveBuffersPerGate);
+
+        return actualExclusiveBuffersPerGate + maxFloatingBuffers;
+    }
```

Comment:
This seems mostly duplicates the calculation for creating `GateBuffersNumCalculator`.